### PR TITLE
v4.0.x: removing an overly aggressive error check in binding

### PR DESCRIPTION
**Pull request by awlauria**
In bind_generic() there's a loop that picks a starting trg_obj and then
walks through a loop of next = trg_obj->next_cousin until it has made
total_cpus assignments.  But the code doesn't accept that those assignments
might not be adjacent objects.

Example:
% mpirun -np 2 --report-bindings --map-by ppr:2:node:pe=3 \
    --cpu-set 4,5,7,8,9,11 -bind-to hwthread:overload-allowed
> MCW 0 : [..../BB.B/..../....]
> MCW 1 : [..../..../BB.B/....]

It will want to assign 3 cpus and will loop through
  trg_obj 00001 (with ncpus 1)
  trg_obj 000001 (with ncpus 1)
  trg_obj 0000001 (with ncpus 0)
  trg_obj 000000011 (with ncpus 1)

The original code on the third entry would see num_bound for the
object become too high for its ncpus and think oversubscription was
happening.  I changed it to only ++num_bound eg to use that object
if the object has cpus in its cpuset after intersected with the
allowed/available masks.

The error message from the original code (if you remove the overload-allowed)
would be
> A request was made to bind to that would result in binding more
> processes than cpus on a resource:
>    Bind to:     HWTHREAD
>    Node:        ...
>    #processes:  1
>    #cpus:      0

bot:notacherrypick

Signed-off-by: Austen Lauria <awlauria@us.ibm.com>

### CODE REVIEW

**Comment by ibm-ompi** [comment]
The IBM CI (GNU/Scale) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/a6ae2e21194512012b9c0d1aa5c6aaa7

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/a553dd241ad3ad5e441a52c058eaf73d

**Comment by awlauria** [comment]
Ugh - this needs another patch. Marking as a draft for now until I get this resolved.

**Comment by ibm-ompi** [comment]
The IBM CI (PGI) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/9a96f3d62454b4e284e7d1544a2ccff8

**Comment by gpaulsen** [comment]
Also need a v4.1.x version of this please (perhaps after others have gone in).

File: orte/mca/rmaps/base/rmaps_base_binding.c [hunk 1]
```diff
@@ -236,7 +239,10 @@ static int bind_generic(orte_job_t *jdata,
                 }
             }
             /* bind the proc here */
-            hwloc_bitmap_or(totalcpuset, totalcpuset, trg_obj->cpuset);
+            // was: hwloc_bitmap_or(totalcpuset, totalcpuset, trg_obj->cpuset);
+            cpus = opal_hwloc_base_get_available_cpus(node->topology->topo, trg_obj);
```
Comment (gpaulsen) [inline orte/mca/rmaps/base/rmaps_base_binding.c:243]: This `opal_hwloc_base_get_available_cpus` looks like the offending call.

Comment (awlauria) [reply orte/mca/rmaps/base/rmaps_base_binding.c:243]: Yeah - it comes from a much larger patch that mark did that will need some eyeballing.

**Comment by rhc54** [comment]
I remember this one from long ago - it created quite a few collateral problems. We eventually fixed it a different way in the master branch to avoid the induced problems. If you really want to resolve this on the release branch, I would recommend backporting what is on master instead of using this patch.

**Comment by hppritcha** [comment]
bot:ompi:retest

**Comment by gpaulsen** [comment]
retest

**Comment by ibm-ompi** [comment]
The IBM CI (XL) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/5d4da738905f5f395586c165f4c1ddee

**Pull request closed without merging**